packages/nodejs/src/transmitter.ts

```diff
@@ -4,79 +4,171 @@ import http from "http"
 
 import { Configuration } from "./config"
 import { URL, URLSearchParams } from "url"
-import { HashMap } from "@appsignal/types"
+
+type TransmitterRequestOptions = {
+  method: string
+  params?: URLSearchParams
+  callback: (stream: http.IncomingMessage) => void
+  onError: (error: Error) => void
+}
 
 export class Transmitter {
   #config: Configuration
   #url: string
-  #data: string
+  #body: string
 
-  constructor(url: string, data = "") {
+  constructor(url: string, body = "") {
     this.#config = new Configuration({})
     this.#url = url
-    this.#data = data
+    this.#body = body
   }
 
-  public async transmit() {
-    return new Promise<HashMap<any>>((resolve, reject) => {
-      const config_data = this.#config.data
-      const params = new URLSearchParams({
-        api_key: config_data["pushApiKey"] || "",
-        name: config_data["name"] || "",
-        environment: config_data["environment"] || "",
-        hostname: config_data["hostname"] || ""
-      })
+  public download() {
```

Comment:
> If we every want to do a "normal GET request" I wonder if we can make this optional by adding a stream: true type of argument here, or move it to a getStream function.

I'm happy to implement some sort of `stream: true`, and that way we can have `.request` take care of the handling of the response to some extent. When looking at the original implementation of the extension download, I drew the line at the separation between the HTTP request and the filesystem -- I left the filesystem handling bits to the extension, and moved the HTTP request bits to the transmitter. In that sense, it's true that `.download` is not a good name, as it doesn't really download anything by itself. I'll apply this change and maybe rename it to `.downloadStream` or `.getStream` as you suggested.

>  Can we return the fully read response to the resolve and write it directly without streaming? It's probably not very efficient, right?

In practice, it wouldn't be a problem if we did. The agent is just a few megabytes. It does feel to me like piping the response stream to the file stream is the idiomatic Node.js thing to do, though.

> I kind of want to rename this to get, but I see we pass through the response body stream to the file streamer. 

I have aimed for `.request` to be the "reasonable enough" interface for making generic HTTP requests, and `.transmit` and `.download` are the specific wrappers for the unique things that our application currently wants to do. Based on that, I'm not so sure about renaming them to `.get` and `.post`, precisely because they don't do normal GET and POST requests, they do somewhat specific things instead:

- `.transmit` is "do a POST request with these four specific values from our config as query parameters, and resolve to the status code and the response value parsed as JSON, discarding it if it's not valid JSON"
- `.download` is a bit more reasonable, but it's still "do a GET request and resolve to the resulting return stream, or reject if the status code is > 400"

Reply:
> I'm happy to implement some sort of stream: true, and that way we can have .request take care of the handling of the response to some extent.

I wouldn't do that now. Only when we come back to this and need to add more types of requests to it. For now I'd rename this function to `downloadStream` to make clear it's not a normal "GET" or download type of request.

> In practice, it wouldn't be a problem if we did. The agent is just a few megabytes. It does feel to me like piping the response stream to the file stream is the idiomatic Node.js thing to do, though.

Yeah let's keep it this way.

Reply:
Renamed it to `.downloadStream`.